sno/sno_repo.py

```diff
@@ -0,0 +1,318 @@
+        # Write working copy config:
+        WorkingCopy.write_config(self, wc_path, bare)
+
+    def activate(self):
+        """
+        We create new+tidy repos in .sno/ but we don't write the .git file pointing to .sno/ until everything
+        else is ready, and until that file is written, git or sno commands won't find the repo.
+        So, if creation fails, the result will be something that doesn't work at all, not something that half
+        works but is also half corrupted.
+        """
+        if not self.is_new_tidy_repo():
+            # Old+bare repos are always activated - since all the files are right there in the root directory,
+            # we can't reveal them by writing the .git file. So, no action is required here.
+            return
+
+        dot_git_path = self.workdir_path / ".git"
+        dot_sno_path = self.gitdir_path
+        # .sno is linked from .git at this point, which means git (or sno) can find it
+        # and so the repository is activated (ie, sno or git commands will work):
+        dot_git_path.write_text("gitdir: .sno\n", encoding="utf-8")
+
+        if is_windows:
+            # Hide .git and .sno
+            # Best effort: if it doesn't work for some reason, continue anyway.
+            subprocess.call(["attrib", "+h", str(dot_git_path)])
+            subprocess.call(["attrib", "+h", str(dot_sno_path)])
+
+    def is_old_bare_repo(self):
+        """Old style sno repos were bare git repos. They were not "tidy": all of the git internals were visible."""
+        return super().is_bare and self.gitdir_path.stem != ".sno"
+
+    def is_new_tidy_repo(self):
+        """New style sno repos are "tidy": they hide the git internals in a ".sno" directory."""
+        return self.gitdir_path.stem == ".sno"
+
+    @property
+    def BARE_CONFIG_KEY(self):
+        return (
+            SnoConfigKeys.SNO_WORKINGCOPY_BARE
+            if self.is_old_bare_repo()
+            else SnoConfigKeys.CORE_BARE
+        )
+
+    @property
+    def version(self):
+        """Returns the sno repository version - eg 2 for 'Datasets V2' See DATASETS_v2.md"""
+        return get_repo_version(self)
+
+    @property
+    def workingcopy_path(self):
+        """Return the path to the sno working copy, if one exists."""
+        repo_cfg = self.config
+        path_key = SnoConfigKeys.SNO_WORKINGCOPY_PATH
+        return Path(repo_cfg[path_key]) if path_key in repo_cfg else None
+
+    @property
+    def is_bare(self):
+        """
+        True if this sno repo is bare - it has no sno working copy.
+        The repo may or may not also be a bare git repository - this is an implementation detail.
+        That information is at super().is_bare
+        """
+        repo_cfg = self.config
+        bare_key = self.BARE_CONFIG_KEY
+        return repo_cfg.get_bool(bare_key) if bare_key in repo_cfg else False
+
+    @contextlib.contextmanager
+    def unlocked_git_index(self):
```

Comment:
> GIT_INDEX_FILE is the path to the index file (non-bare repositories only).


might be worth using this to change where git subcommands looks for the index, rather than deleting & re-creating it. That will make this safe against concurrent runs and crashes

Reply:
Done! Not sure how I missed this - it was the kind of thing I was hoping to do